update device mesh to include slicing (#2958)
pytorch/tutorials: recipes_source/distributed_device_mesh.rst

    torchrun --nproc_per_node=8 hsdp.py

How to use DeviceMesh for your custom parallel solutions
--------------------------------------------------------

> When working with large scale training, one can have a more complicated custom parallel training composition. For example, one may need to slice out submeshes for different parallelism solutions.

Reviewer: ```suggestion
When working with large scale training, you might have more complex custom parallel training composition. For example, you may need to slice out sub-meshes for different parallelism solutions.
```

Reviewer: If this is for 2.4, can you please label with the 2.4 label.

Author: @svekars Not sure why but it seems I can't add reviewer and labels. I was gonna add you and one team member @wanchaol as the reviewer.

DeviceMesh allows users to slice child mesh from the parent mesh and re-use the NCCL communicators already created when the parent mesh is initialized.

.. code-block:: python

    from torch.distributed.device_mesh import init_device_mesh
    mesh_3d = init_device_mesh("cuda", (2, 2, 2), mesh_dim_names=("replicate", "shard", "tp"))

    # Users can slice child meshes from the parent mesh.
    hsdp_mesh = mesh_3d["replicate", "shard"]
    tp_mesh = mesh_3d["tp"]

    # Users can access the underlying process group thru `get_group` API.

>     replicate_group = hsdp_mesh["Replicate"].get_group()

Reviewer 2: these strs need to be lower case
